packages/net/http/payload.pony

```diff
@@ -56,187 +117,248 @@ class iso Payload
-      h(consume this, Payload.response(StatusInternalServerError))
-    end
+None /*    try
+      (session as HTTPSession tag)(
+         Payload.response(StatusInternalServerError))
+    end */
 
-  fun _write(conn: TCPConnection, keepalive: Bool = true) =>
+  fun val _write( keepalive: Bool = true, conn: TCPConnection tag ) =>
     """
-    Writes the payload to a TCP connection.
+    Writes the payload to an HTTPSession.  Requests and Responses differ
+    only in the first line of text - everything after that is the same
+    format.
     """
     if _response then
-      _write_response(conn, keepalive)
+      _write_response(keepalive, conn)
     else
-      _write_request(conn, keepalive)
+      _write_request(keepalive, conn)
     end
 
-  fun _write_request(conn: TCPConnection, keepalive: Bool) =>
+    _write_common( conn )
+
+  fun val _write_request( keepalive: Bool, conn: TCPConnection tag) =>
     """
-    Writes an an HTTP request.
+    Writes the 'request' parts of an HTTP message.
     """
-    let len = 15 + (4 * _headers.size()) + (4 * _body.size())
-    var list = recover Array[ByteSeq](len) end
-
-    list.push(method)
-    list.push(" ")
-    list.push(url.path)
+    conn.write(method + " " + url.path)
 
     if url.query.size() > 0 then
-      list.push("?")
-      list.push(url.query)
-    end
+      conn.write( "?" + url.query)
+      end
 
     if url.fragment.size() > 0 then
-      list.push("#")
-      list.push(url.fragment)
-    end
+      conn.write( "#" + url.fragment)
+      end
 
-    list.push(" ")
-    list.push(proto)
+    conn.write( " " + proto + "\r\n" )
 
     if not keepalive then
-      list.push("\r\nConnection: close")
-    end
-
-    list.push("\r\nHost: ")
-    list.push(url.host)
-    list.push(":")
-    list.push(url.port.string())
-    // TODO: basic authorization header
-
-    list = _add_headers(consume list)
-    list = _add_body(consume list)
-
-    conn.writev(consume list)
+      conn.write("Connection: close\r\n")
+      end
 
-  fun _write_response(conn: TCPConnection, keepalive: Bool) =>
+    conn.write("Host: " + url.host + ":" + url.port.string() + "\r\n")
+  
+  fun val _write_common( conn: TCPConnection tag ) =>
     """
-    Write as an HTTP response.
+    Writes the parts of an HTTP message common to both requests and
+    responses.
     """
-    let len = 8 + (4 * _headers.size()) + (4 * _body.size())
-    var list = recover Array[ByteSeq](len) end
-
-    list.push(proto)
-    list.push(" ")
-    list.push(method)
+    _write_headers( conn )
 
+    // In oneshot mode we send the entire stored body.
+    if transfer_mode is OneshotTransfer then
+      for piece in _body.values() do
+        conn.write(piece)
+        end
+      end
+          
+  fun val _write_response( keepalive: Bool, conn: TCPConnection tag ) =>
+    """
+    Write the response-specific parts of an HTTP message.
+    """
+    conn.write(proto + " " + status.string() + " " + method + "\r\n")
```

Comment:
String concatenation using `+` here can be slow since it allocates a new string on every `+`